src/platform/ESP32/ESP32FactoryDataProvider.cpp

```diff
@@ -0,0 +1,161 @@
+{
+    static constexpr size_t kSpake2pSerializedVerifier_MaxBase64Len =
+        BASE64_ENCODED_LEN(chip::Crypto::kSpake2p_VerifierSerialized_Length) + 1;
+
+    CHIP_ERROR err                                            = CHIP_NO_ERROR;
+    char verifierB64[kSpake2pSerializedVerifier_MaxBase64Len] = { 0 };
+    size_t verifierB64Len                                     = 0;
+
+    err =
+        ESP32Config::ReadConfigValueStr(ESP32Config::kConfigKey_Spake2pVerifier, verifierB64, sizeof(verifierB64), verifierB64Len);
+    ReturnErrorOnFailure(err);
+
+    verifierLen = chip::Base64Decode32(verifierB64, verifierB64Len, reinterpret_cast<uint8_t *>(verifierB64));
+    ReturnErrorCodeIf(verifierLen > verifierBuf.size(), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    memcpy(verifierBuf.data(), verifierB64, verifierLen);
+    verifierBuf.reduce_size(verifierLen);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetCertificationDeclaration(MutableByteSpan & outBuffer)
+{
+    size_t certSize;
+    ReturnErrorOnFailure(
+        ESP32Config::ReadConfigValueBin(ESP32Config::kConfigKey_CertDeclaration, outBuffer.data(), outBuffer.size(), certSize));
+    outBuffer.reduce_size(certSize);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetFirmwareInformation(MutableByteSpan & out_firmware_info_buffer)
+{
+    // TODO: We need a real example FirmwareInformation to be populated.
+    out_firmware_info_buffer.reduce_size(0);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetDeviceAttestationCert(MutableByteSpan & outBuffer)
+{
+    size_t certSize;
+    ReturnErrorOnFailure(
+        ESP32Config::ReadConfigValueBin(ESP32Config::kConfigKey_DACCert, outBuffer.data(), outBuffer.size(), certSize));
+    outBuffer.reduce_size(certSize);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetProductAttestationIntermediateCert(MutableByteSpan & outBuffer)
+{
+    size_t certSize;
+    ReturnErrorOnFailure(
+        ESP32Config::ReadConfigValueBin(ESP32Config::kConfigKey_PAICert, outBuffer.data(), outBuffer.size(), certSize));
+    outBuffer.reduce_size(certSize);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::SignWithDeviceAttestationKey(const ByteSpan & digestToSign, MutableByteSpan & outSignBuffer)
+{
+    Crypto::P256ECDSASignature signature;
+    Crypto::P256Keypair keypair;
+
+    VerifyOrReturnError(IsSpanUsable(outSignBuffer), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(IsSpanUsable(digestToSign), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(outSignBuffer.size() >= signature.Capacity(), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    uint8_t privKeyBuf[kDACPrivateKeySize];
+    uint8_t pubKeyBuf[kDACPublicKeySize];
+    size_t privKeyLen = sizeof(privKeyBuf);
+    size_t pubKeyLen  = sizeof(pubKeyBuf);
+
+    ReturnErrorOnFailure(
+        ESP32Config::ReadConfigValueBin(ESP32Config::kConfigKey_DACPrivateKey, privKeyBuf, privKeyLen, privKeyLen));
```

Comment:
Is the storage holding the private key encrypted? If not, it may be a problem. Can that partition be encrypted with a per-device key that cannot be leaked by an attacker?

Reply:
In the examples storage(NVS) holding the private key is not encrypted but, can be encrypted with per-device key. Encryption key is stored in a different place which is protected using Flash encryption.